### PR TITLE
feat: stringify dataset column contents during data loading

### PR DESCRIPTION
**Pull request by danielezhu**
*Issue #, if available:*

*Description of changes:*
Currently, the JSON parsing code extracts raw data from the input dataset using user-provided JMESPath expressions. This raw data can be of any data type, such as booleans, floats, etc.

Certain columns (e.g. target output) should _always_ be strings, even if the raw data from the dataset is not in string form. For example, the dataset could contain target outputs that are booleans, but due to the nature of the evaluation algorithms, the target output needs to be something like `"True"` or `"False"`.

This PR adds a function `cast_to_string`, and updates the `_parse_column` method to call `cast_to_string` if appropriate. Additionally, this PR changes `class ColumnNames(Enum)` in `constants.py` to `class DatasetColumns(Enum)`, where instead of enumerating just the _names_ of the columns, it enumerates `Column` objects, where `Column` is a new class that I introduce.

`Column` contains a `name` attribute which represents the original strings that were being enumerated by `class ColumnNames(Enum)` and a `should_cast` attribute which represents whether the contents of this column should be casted to string during data loading.

The reason I added this class is so that we don't create a separate list for tracking which columns should/shouldn't be casted. Doing so would require updating this list manually whenever new column types are introduced, which is a source of human error that should be avoided. PR #171 was raised precisely because we were doing something similar in the past.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by xiaoyi-cheng — APPROVED** [review]

**Review by keerthanvasist — APPROVED** [review]

**Comment by lucfra** [comment]
Thanks @danielezhu ! Just got a question. Will this convert any field of the dataset into a string? 
If so, I would strongly recommend against it.
We should only convert fields that are _required_ to be strings by the eval algorithms (most notably input and targets, if present).

**Review by lucfra — CHANGES_REQUESTED** [review]
Thanks @danielezhu ! Just got a question. Will this convert any field of the dataset into a string?
If so, I would strongly recommend against it.
We should only convert fields that are required to be strings by the eval algorithms (most notably input and targets, if present).

**Review by lucfra — CHANGES_REQUESTED** [review]
Thanks @danielezhu ! Just got a question. Will this convert any field of the dataset into a string?
If so, I would strongly recommend against it.
We should only convert fields that are required to be strings by the eval algorithms (most notably input and targets, if present).

**Review by franluca — DISMISSED** [review]
Thanks @danielezhu ! Just got a question. Will this convert any field of the dataset into a string?
If so, I would strongly recommend against it.
We should only convert fields that are required to be strings by the eval algorithms (most notably input and targets, if present).

**Comment by danielezhu** [comment]
Luca raised a good point about not casting _every_ column. It is clear why we shouldn't do this if we consider the log probability columns. I will update this PR after Luca sends me the list of columns that should be converted to strings, and those that should not.

**Comment by lucfra** [comment]
However I have another question. Can in principle be the fields also lists? like having a list of target output(s) ? Or do we prohibit this?

**Review by xiaoyi-cheng — DISMISSED** [review]

File: src/fmeval/data_loaders/json_parser.py [hunk 1]
```diff
@@ -55,34 +60,35 @@ def __init__(
...
 
         :param config: see DataConfig docstring
         """
         self._parsers = dict()
 
         for attribute_name, attribute_value in config.__dict__.items():
             if attribute_name.endswith(DATA_CONFIG_LOCATION_SUFFIX) and attribute_value is not None:
-                key = attribute_name.replace(DATA_CONFIG_LOCATION_SUFFIX, "")
+                column_name = attribute_name.replace(DATA_CONFIG_LOCATION_SUFFIX, "")
                 assert_condition(
```
[the remark this answers is not in the document]
Comment (danielezhu) [reply src/fmeval/data_loaders/json_parser.py:82]: I think this logic is best explained in the docstring `__init__`.

File: src/fmeval/data_loaders/json_parser.py [hunk 2]
```diff
@@ -206,3 +215,35 @@ def _validate_parsed_columns_lengths(parsed_columns_dict: Dict[str, List[Any]]):
...
+            if args.dataset_mime_type == MIME_TYPE_JSON:
+                return [str(x) for x in result]
+            elif args.dataset_mime_type == MIME_TYPE_JSONLINES:
+                return str(result)
+            else:
+                raise EvalAlgorithmInternalError(  # pragma: no cover
+                    f"args.dataset_mime_type is {args.dataset_mime_type}, but only JSON and JSON Lines are supported."
+                )
+        except Exception:
+            raise EvalAlgorithmClientError(
```
[the remark this answers is not in the document]
Comment (danielezhu) [reply src/fmeval/data_loaders/json_parser.py:245]: Done

**Review by xiaoyi-cheng — APPROVED** [review]

**Review by oyangz — APPROVED** [review]

**Pull request merged**